hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractListenerMessageTask.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.impl.protocol.task;
+
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.instance.impl.Node;
+import com.hazelcast.internal.nio.Connection;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+
+/**
+ * Base message task for listener registration tasks
+ */
+public abstract class AbstractListenerMessageTask<P, T>
```

Comment:
This class has no specific logic to Listeners. It becomes more clear when you make `eventRegistrationFuture`  a local variable. I think removing it will make both AbstractAddListenerMessageTask and AbstractRemoveListenerMessageTask.

Reply:
Since it has some simplification because of generics. We may rename it to something more generic. 
`AbstractAsyncMessageTask` ?
AbstractAddress and AbstractInvocationMessageTask can extend this class. I see similar pattern there also. 
Also some usages has optimizations while others are missing. 
See 

```
 @Override
    public void execute(Runnable command) {
        if (Thread.currentThread().getClass() == PartitionOperationThread.class) {
            // instead of offloading it to another thread, we run on the partition thread. This will speed up throughput.
            command.run();
        } else {
            ExecutionService executionService = nodeEngine.getExecutionService();
            Executor executor = executionService.getExecutor(ExecutionService.ASYNC_EXECUTOR);
            executor.execute(command);
        }
    }

    @Override
    public void accept(Object response, Throwable throwable) {
        if (throwable == null) {
            beforeResponse();
            sendResponse(response);
            afterResponse();
        } else {
            beforeResponse();
            handleProcessingFailure(throwable);
            afterResponse();
        }
    }
```